airbyte-workers/src/test/java/io/airbyte/workers/general/DefaultReplicationWorkerTest.java

```diff
@@ -292,6 +294,33 @@ void testReplicationRunnableWorkerFailure() throws Exception {
         .anyMatch(f -> f.getFailureOrigin().equals(FailureOrigin.REPLICATION) && f.getStacktrace().contains(WORKER_ERROR_MESSAGE)));
   }
 
+  @Test
+  void testOnlyStateAndRecordMessagesDeliveredToDestination() throws Exception {
+    final AirbyteMessage LOG_MESSAGE = AirbyteMessageUtils.createLogMessage(Level.INFO, "a log message");
+    final AirbyteMessage TRACE_MESSAGE = AirbyteMessageUtils.createTraceMessage("a trace message", 123456.0);
+    when(source.attemptRead()).thenReturn(Optional.of(RECORD_MESSAGE1), Optional.of(LOG_MESSAGE), Optional.of(TRACE_MESSAGE),
+        Optional.of(RECORD_MESSAGE2), Optional.empty());
+
+    final ReplicationWorker worker = new DefaultReplicationWorker(
+        JOB_ID,
+        JOB_ATTEMPT,
+        source,
+        mapper,
+        destination,
+        messageTracker,
+        recordSchemaValidator,
+        workerMetricReporter);
+
+    worker.run(syncInput, jobRoot);
+
+    verify(source).start(sourceConfig, jobRoot);
+    verify(destination).start(destinationConfig, jobRoot);
+    verify(destination).accept(RECORD_MESSAGE1);
+    verify(destination).accept(RECORD_MESSAGE2);
+    verify(destination, never()).accept(LOG_MESSAGE);
+    verify(destination, never()).accept(TRACE_MESSAGE);
+  }
```

Comment:
@lmossman can you help me understand why this test is failing?  I'm not building the `source.attemptRead` mock properly...

Reply:
@evantahler this doesn't fully fix the test, but one error I saw when I ran this locally was an NPE being thrown when the DefaultReplicationWorker tries to call `getType()` on the message, and that is because the `mapper` was not properly being mocked to return the LOG and TRACE messages back. I added these two lines after line 300 of your test and it got rid of that NPE:
```
when(mapper.mapMessage(LOG_MESSAGE)).thenReturn(LOG_MESSAGE);
when(mapper.mapMessage(TRACE_MESSAGE)).thenReturn(TRACE_MESSAGE);
```
Though maybe we should just be mocking that `mapMessage()` method in the `mapper` to just return whatever it is given as input

Reply:
So now the test is failing because it doesn't see the second RECORD message ever get passed to the destination.accept()

I tried adding logging right above the `if (message.getType() == Type.RECORD || message.getType() == Type.STATE)` line in DefaultReplicationWorker, and it seems like it is only ever seeing the first record message, then the log, then the trace, and it never sees the second message before the threads close. Not sure why

Reply:
@evantahler Ah, I understand why this is happening. It's because of this mock of `source.isFinished()`: https://github.com/airbytehq/airbyte/blob/acbdac326b1b02977130361585150ec079350913/airbyte-workers/src/test/java/io/airbyte/workers/general/DefaultReplicationWorkerTest.java#L125

This makes `source.isFinished()` return `true` the fourth time it is called, so the source will only emit 3 messages before closing. You'll need to modify that mock in your test to match your test behavior to get this to work.

Reply:
AH HA! Thank you @lmossman!